### ISSUE TITLE
Unexpected result from d2dtf() very near day boundary in UTC

### ISSUE THREAD

**Issue by taldcroft** [liberfa/erfa#76]
It seems that for a time *just* before the day boundary, it's possible to get the wrong day / hour / min / second representation:
```
In [11]: t = Time('1960-01-02') + 0.0

In [12]: t.jd1
Out[12]: 2436935.0

In [13]: t.jd2
Out[13]: 0.4999999999999998

In [14]: erfa.d2dtf('TAI', 6,  t.jd1, t.jd2)
Out[14]: (1960, 1, 2, (0, 0, 0, 0))

In [15]: erfa.d2dtf('UTC', 6,  t.jd1, t.jd2)
Out[15]: (1960, 1, 1, (23, 59, 60, 0))
```
This doesn't seem to occur for more "modern" dates like 1972-01-02.

**Comment by mhvk** [comment]
I think pre-1972 and per leap seconds, the clocks were allowed to drift; see
https://github.com/liberfa/erfa/blob/55fbff2fc02debfeb9a0bba61fcc58513ca29ce6/src/dat.c#L239-L240

I think given that drift, it would be unavoidable to sometimes get 60 seconds (but it would not be a full second).

**Comment by taldcroft** [comment]
I thought that while the UTC scale was allowed to drift that the UTC representation of the time was, by definition, always having seconds < 60. In other words there was not yet this idea of letting there be more than 60 UTC seconds in a minute.

But in truth I'm not an expert here, especially on the historical side. Here is a different demonstration of what seems unexpected to me:
```
In [22]: t                                                                                                                                         
Out[22]: <Time object: scale='utc' format='iso' value=1960-01-03 00:00:00.000>

In [23]: erfa.d2dtf('UTC', 6,  t.jd1, t.jd2)                                                                                                       
Out[23]: (1960, 1, 3, (0, 0, 0, 0))

In [24]: erfa.d2dtf('UTC', 6,  t.jd1, t.jd2 - 20e-12/86400)                                                                                         
Out[24]: (1960, 1, 2, (23, 59, 60, 0))

In [25]: t.jd1                                                                                                                                     
Out[25]: 2436936.0

In [26]: t.jd2                                                                                                                                     
Out[26]: 0.5
```

Clearly when you subtract 20 picosec from the midnight, the year/mon/day hr/min/sec representation to the nearest microsec should not change at all. Instead it is rounding 59.999999999 to 60.0 but not recognizing that this should then wrap to 1960-01-03 00:00:00.000000.

**Comment by timj** [comment]
Doesn't UTC require that a minute can have between 59 and 61 seconds in it? So if there is a leap second added that does mean that the seconds will indicate 60.

See for example https://www.iausofa.org/2021_0125_C/sofa/sofa_ts_c.pdf page 14.

**Comment by mhvk** [comment]
Hmm, I think you are right that there is a bug: the code dealing with times beyond 24 hours considers any jump a leap second, however small: https://github.com/liberfa/erfa/blob/55fbff2fc02debfeb9a0bba61fcc58513ca29ce6/src/d2dtf.c#L121-L129

Here `leap` is a boolean for `dleap != 0`. My sense is that the comment "is this a leap-second day" tells that the intention is to deal with regular (mini) leap seconds, not ones as small as `dleap=1.11022e-16`, which is what I find by inserting a print statement. Especially since this equals 2**-53, suspiciously close to what you get for a rounding error for subtracting 0.5 from something that is just one bit off. (`np.finfo(np.float).eps` is 2.220446049250313e-16).

**Comment by mhvk** [comment]
I've sent an e-mail to Patrick Wallace to ask about this...

**Comment by taldcroft** [comment]
Great sleuthing @mhvk!

**Comment by mhvk** [comment]
Just to close the loop here - after puzzlement on all sides, the conclusion is that this is indeed a SOFA bug, which will be fixed in the next SOFA release (possibly as a bug-fix release, TBD). It is nice to see how bugs found via astropy continue to help upstream too! Good that you realized things were unexpected, @taldcroft!

**Comment by mhvk** [comment]
This is fixed in SOFA 17a, released today.

**Comment by mhvk** [comment]
fixed by the update to erfa 1.7.3
